README.rst
Python Client for Cloud Billing Budget API
=====================================================

|beta| |pypi| |versions|

`Cloud Billing Budget API`_: The Cloud Billing Budget API stores Cloud Billing budgets, which define a
budget plan and the rules to execute as spend is tracked against that
plan.

- `Client Library Documentation`_
- `Product Documentation`_

.. |beta| image:: https://img.shields.io/badge/support-beta-orange.svg
   :target: https://github.com/googleapis/google-cloud-python/blob/master/README.rst#beta-support
.. |pypi| image:: https://img.shields.io/pypi/v/google-cloud-billing-budgets.svg
   :target: https://pypi.org/project/google-cloud-billing-budgets
.. |versions| image:: https://img.shields.io/pypi/pyversions/google-cloud-billing-budgets.svg
   :target: https://pypi.org/project/google-cloud-billing-budgets/
.. _Cloud Billing Budget API: https://cloud.google.com/billing/docs/how-to/budget-api-overview
.. _Client Library Documentation: https://googleapis.dev/python/billing-budgets/latest
.. _Product Documentation:  https://cloud.google.com/billing/docs/how-to/budget-api-overview

Quick Start
-----------

In order to use this library, you first need to go through the following steps:

1. `Select or create a Cloud Platform project.`_
2. `Enable billing for your project.`_
3. `Enable the Cloud Billing Budget API.`_
4. `Setup Authentication.`_

.. _Select or create a Cloud Platform project.: https://console.cloud.google.com/project
.. _Enable billing for your project.: https://cloud.google.com/billing/docs/how-to/modify-project#enable_billing_for_a_project
.. _Enable the Cloud Billing Budget API.:  https://cloud.google.com/billing/docs/how-to/budget-api-overview
.. _Setup Authentication.: https://googleapis.dev/python/google-api-core/latest/auth.html

Installation
~~~~~~~~~~~~

Install this library in a `virtualenv`_ using pip. `virtualenv`_ is a tool to
create isolated Python environments. The basic problem it addresses is one of
dependencies and versions, and indirectly permissions.

With `virtualenv`_, it's possible to install this library without needing system
install permissions, and without clashing with the installed system
dependencies.

.. _`virtualenv`: https://virtualenv.pypa.io/en/latest/

Supported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^
Python >= 3.6

Deprecated Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^^
Python == 2.7. Python 2.7 support will be removed on January 1, 2020.

The last version of this library compatible with Python 2.7 is google-cloud-billing-budgets==0.4.0.

Mac/Linux
^^^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    source <your-env>/bin/activate
    <your-env>/bin/pip install google-cloud-billing-budgets


Windows
^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    <your-env>\Scripts\activate
    <your-env>\Scripts\pip.exe install google-cloud-billing-budgets

Next Steps
~~~~~~~~~~

-  Read the `Client Library Documentation`_ for Cloud Billing Budget API
   API to see other available methods on the client.
-  Read the `Cloud Billing Budget API Product documentation`_ to learn
   more about the product and see How-to Guides.
-  View this `README`_ to see the full list of Cloud
   APIs that we cover.

.. _Cloud Billing Budget API Product documentation:  https://cloud.google.com/billing/docs/how-to/budget-api-overview
.. _README: https://github.com/googleapis/google-cloud-python/blob/master/README.rst


CHANGELOG.md
# Changelog
## [0.4.2](https://github.com/googleapis/python-billingbudgets/compare/v0.4.1...v0.4.2) (2022-06-08)


### Bug Fixes

* **deps:** require protobuf<4.0.0 on v0 branch ([#213](https://github.com/googleapis/python-billingbudgets/issues/213)) ([f46e868](https://github.com/googleapis/python-billingbudgets/commit/f46e868a980b67a655c756d2897f81a7f4526f61))

### [0.4.1](https://github.com/googleapis/python-billingbudgets/compare/v0.4.0...v0.4.1) (2022-04-01)


### Bug Fixes

* **deps:** require google-api-core >= 1.31.5, >= 2.3.2 on v0 release ([#186](https://github.com/googleapis/python-billingbudgets/issues/186)) ([b003202](https://github.com/googleapis/python-billingbudgets/commit/b0032029a71ead1dd78affb29b09bd8ac0caf370))

## [0.4.0](https://www.github.com/googleapis/python-billingbudgets/compare/v0.3.0...v0.4.0) (2020-05-18)


### Features

* refreshes beta release ([#28](https://www.github.com/googleapis/python-billingbudgets/issues/28)) ([49b6e2a](https://www.github.com/googleapis/python-billingbudgets/commit/49b6e2a22a207e2a0c8dfc34dce7df7891003156))

## [0.3.0](https://www.github.com/googleapis/python-billingbudgets/compare/v0.2.0...v0.3.0) (2020-03-18)


### Features

* release to beta ([#15](https://www.github.com/googleapis/python-billingbudgets/issues/15)) ([b7c3605](https://www.github.com/googleapis/python-billingbudgets/commit/b7c36056d166fef72aacc80773868c447bf8c1b4))

## 0.2.0

12-12-2019 12:36 PST


### New Features
- Deprecate resource name helper methods (via synth). ([#9830](https://github.com/googleapis/google-cloud-python/pull/9830))

### Internal / Testing Changes
- Add 3.8 unit tests (via synth). ([#9933](https://github.com/googleapis/google-cloud-python/pull/9933))
- Remove TODOs in proto comments (via synth). ([#9912](https://github.com/googleapis/google-cloud-python/pull/9912))
- Add comments to proto files (via synth). ([#9854](https://github.com/googleapis/google-cloud-python/pull/9854))

## 0.1.0

11-15-2019 10:34 PST

### New Features
- Initial generation of billing budget. ([#9622](https://github.com/googleapis/google-cloud-python/pull/9622))

### Documentation
- Add Python 2 sunset banner (via synth). ([#9813](https://github.com/googleapis/google-cloud-python/pull/9813))


owl-bot-staging/v1beta1/noxfile.py
# -*- coding: utf-8 -*-
# Copyright 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import os
import pathlib
import shutil
import subprocess
import sys


import nox  # type: ignore

ALL_PYTHON = [
    "3.6",
    "3.7",
    "3.8",
    "3.9",
    "3.10",
]

CURRENT_DIRECTORY = pathlib.Path(__file__).parent.absolute()

LOWER_BOUND_CONSTRAINTS_FILE = CURRENT_DIRECTORY / "constraints.txt"
PACKAGE_NAME = subprocess.check_output([sys.executable, "setup.py", "--name"], encoding="utf-8")

BLACK_VERSION = "black==19.10b0"
BLACK_PATHS = ["docs", "google", "tests", "samples", "noxfile.py", "setup.py"]
DEFAULT_PYTHON_VERSION = "3.9"

nox.sessions = [
    "unit",
    "cover",
    "mypy",
    "check_lower_bounds"
    # exclude update_lower_bounds from default
    "docs",
    "blacken",
    "lint",
    "lint_setup_py",
]

@nox.session(python=ALL_PYTHON)
def unit(session):
    """Run the unit test suite."""

    session.install('coverage', 'pytest', 'pytest-cov', 'pytest-asyncio', 'asyncmock; python_version < "3.8"')
    session.install('-e', '.')

    session.run(
        'py.test',
        '--quiet',
        '--cov=google/cloud/billing/budgets_v1beta1/',
        '--cov=tests/',
        '--cov-config=.coveragerc',
        '--cov-report=term',
        '--cov-report=html',
        os.path.join('tests', 'unit', ''.join(session.posargs))
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def cover(session):
    """Run the final coverage report.
    This outputs the coverage report aggregating coverage from the unit
    test runs (not system test runs), and then erases coverage data.
    """
    session.install("coverage", "pytest-cov")
    session.run("coverage", "report", "--show-missing", "--fail-under=100")

    session.run("coverage", "erase")


@nox.session(python=ALL_PYTHON)
def mypy(session):
    """Run the type checker."""
    session.install('mypy', 'types-pkg_resources')
    session.install('.')
    session.run(
        'mypy',
        '--explicit-package-bases',
        'google',
    )


@nox.session
def update_lower_bounds(session):
    """Update lower bounds in constraints.txt to match setup.py"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'update',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )


@nox.session
def check_lower_bounds(session):
    """Check lower bounds in setup.py are reflected in constraints file"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'check',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )

@nox.session(python=DEFAULT_PYTHON_VERSION)
def docs(session):
    """Build the docs for this library."""

    session.install("-e", ".")
    session.install("sphinx==4.0.1", "alabaster", "recommonmark")

    shutil.rmtree(os.path.join("docs", "_build"), ignore_errors=True)
    session.run(
        "sphinx-build",
        "-W",  # warnings as errors
        "-T",  # show full traceback on exception
        "-N",  # no colors
        "-b",
        "html",
        "-d",
        os.path.join("docs", "_build", "doctrees", ""),
        os.path.join("docs", ""),
        os.path.join("docs", "_build", "html", ""),
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint(session):
    """Run linters.

    Returns a failure if the linters find linting errors or sufficiently
    serious code quality issues.
    """
    session.install("flake8", BLACK_VERSION)
    session.run(
        "black",
        "--check",
        *BLACK_PATHS,
    )
    session.run("flake8", "google", "tests", "samples")


@nox.session(python=DEFAULT_PYTHON_VERSION)
def blacken(session):
    """Run black. Format code to uniform standard."""
    session.install(BLACK_VERSION)
    session.run(
        "black",
        *BLACK_PATHS,
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint_setup_py(session):
    """Verify that setup.py is valid (including RST check)."""
    session.install("docutils", "pygments")
    session.run("python", "setup.py", "check", "--restructuredtext", "--strict")
